File: tinyml/cc-coding/src/cpp/virtual.cc
#include <iostream>
using namespace std; // bad habits

class A {
public:
    int forward(int x) const {
        return x;
    }
};

class B : public A {
public:
    int forward(int x) const {
        return 2*x;
    }
};

void print_res(const A* model, int x) {
    cout << model->forward(x) << endl;
}

int main() {
    B* m = new B();
    print_res(m, 3); // 3
    delete m;
}

File: tinyml/cc-coding/src/cpp/virtual.py
class A:
    def forward(self, x):
        return x

class B(A):
    def forward(self, x):
        return 2 * x

def print_res(model: A, x: int):
    print(model.forward(x))

if __name__ == "__main__":
    m = B()
    print_res(m, 3)  # 6